# Capstone Notebook

## Introduction

This is a notebook that will be used for the Applied Data Science Capstone course from coursera.

In [1]:
import pandas as pd
import numpy as np

In [2]:
print('Hello Capstone Project Course!')

Hello Capstone Project Course!


# Segmenting and Clustering Neighborhoods in Toronto
## Week 3 Peer Reviewed Assignment

## Data retrieval

In [1]:
import requests
import lxml.html as lh
import pandas as pd

In [12]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#Create a handle, page, to handle the contents of the website
page = requests.get(url)
#Store the contents of the website under doc
doc = lh.fromstring(page.content)
#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//*[@id="mw-content-text"]/div/table[1]//tr')

In [13]:
#Check the length of all rows
print(len(tr_elements))
print([len(T) for T in tr_elements[:]])

181
[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]


**Now that we have the info from the table, we convert it into a dataframe**

In [46]:
scrapped_dict = {}
headers = []
for header in tr_elements[0]:
    scrapped_dict[header.text_content().replace('\n', '')] = []
    headers.append(header.text_content().replace('\n', ''))
for t in tr_elements[1:]:
    for i, col in enumerate(t):
        scrapped_dict[headers[i]].append(col.text_content().replace('\n', ''))

In [47]:
print(scrapped_dict)

{'Postal code': ['M1A', 'M2A', 'M3A', 'M4A', 'M5A', 'M6A', 'M7A', 'M8A', 'M9A', 'M1B', 'M2B', 'M3B', 'M4B', 'M5B', 'M6B', 'M7B', 'M8B', 'M9B', 'M1C', 'M2C', 'M3C', 'M4C', 'M5C', 'M6C', 'M7C', 'M8C', 'M9C', 'M1E', 'M2E', 'M3E', 'M4E', 'M5E', 'M6E', 'M7E', 'M8E', 'M9E', 'M1G', 'M2G', 'M3G', 'M4G', 'M5G', 'M6G', 'M7G', 'M8G', 'M9G', 'M1H', 'M2H', 'M3H', 'M4H', 'M5H', 'M6H', 'M7H', 'M8H', 'M9H', 'M1J', 'M2J', 'M3J', 'M4J', 'M5J', 'M6J', 'M7J', 'M8J', 'M9J', 'M1K', 'M2K', 'M3K', 'M4K', 'M5K', 'M6K', 'M7K', 'M8K', 'M9K', 'M1L', 'M2L', 'M3L', 'M4L', 'M5L', 'M6L', 'M7L', 'M8L', 'M9L', 'M1M', 'M2M', 'M3M', 'M4M', 'M5M', 'M6M', 'M7M', 'M8M', 'M9M', 'M1N', 'M2N', 'M3N', 'M4N', 'M5N', 'M6N', 'M7N', 'M8N', 'M9N', 'M1P', 'M2P', 'M3P', 'M4P', 'M5P', 'M6P', 'M7P', 'M8P', 'M9P', 'M1R', 'M2R', 'M3R', 'M4R', 'M5R', 'M6R', 'M7R', 'M8R', 'M9R', 'M1S', 'M2S', 'M3S', 'M4S', 'M5S', 'M6S', 'M7S', 'M8S', 'M9S', 'M1T', 'M2T', 'M3T', 'M4T', 'M5T', 'M6T', 'M7T', 'M8T', 'M9T', 'M1V', 'M2V', 'M3V', 'M4V', 'M5V', 'M6

In [48]:
# Check if postal codes are in at most 1 row
print(len(scrapped_dict['Postal code']), len(set(scrapped_dict['Postal code'])))

180 180


In [87]:
# Convert to dataframe
df = pd.DataFrame.from_dict(scrapped_dict)
df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                            
1         M2A      Not assigned                            
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

In [88]:
# Delete Not assigned boroughs and separate the neighborhoods by commas instead of slash
not_assinged_indexes = df[df['Borough'] == 'Not assigned'].index
df.drop(not_assinged_indexes, inplace=True)
df.head()

Postal code           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [89]:
# Check to see how many unassigned boroughs there were
df.shape

(103, 3)

**There were 77 deletions, which matches the unassigned (ctrl + f) in wikipedia**

In [90]:
# Use comma separated neighborhoods
df['Neighborhood'] = df['Neighborhood'].str.replace(' / ', ', ')
df.head()

Postal code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [91]:
# Check for unassigned neighborhoods
df[df['Neighborhood'] == '']

Empty DataFrame
Columns: [Postal code, Borough, Neighborhood]
Index: []

**We can observe that there are no unassigned neighborhoods**

In [92]:
df.shape

(103, 3)

**We now have our data ready for processing.  
shape: (103, 3)**

In [156]:
# Check amount of neighborhoods
suma = 0
for i, neighborhoods in enumerate(df['Neighborhood']):
    n = neighborhoods.split(', ')
    suma += len(n)
print(suma)

209


In [157]:
# Divide neighborhoods in different rows
df_divided = pd.DataFrame(columns=['Postal code', 'Borough', 'Neighborhood'])
for index, row in df.iterrows():
    neighs = row['Neighborhood'].split(',')
    if len(neighs) == 1:
        df_divided = df_divided.append({'Postal code': row['Postal code'], 'Borough': row['Borough'], 'Neighborhood': row['Neighborhood']}, ignore_index=True)
    else:
        for n in neighs:
            df_divided = df_divided.append({'Postal code': row['Postal code'], 'Borough': row['Borough'], 'Neighborhood': n}, ignore_index=True)

In [158]:
df_divided.shape

(209, 3)

**We notice that the shape corresponds to the 209 neighborhoods**

**We cannot get coordinates for every neighbourhood, so we use coordinates for postal codes only**  
The division we did is not used later on

In [159]:
df.head()

Postal code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [160]:
geo_df = pd.read_csv('https://cocl.us/Geospatial_data')
df_joined = pd.merge(df, geo_df, how='left', left_on='Postal code', right_on='Postal Code')

In [161]:
df_joined.head()

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

  Postal Code   Latitude  Longitude  
0         M3A  43.753259 -79.329656  
1         M4A  43.725882 -79.315572  
2         M5A  43.654260 -79.360636  
3         M6A  43.718518 -79.464763  
4         M7A  43.662301 -79.389494

In [168]:
# Drop Postal Code duplicate field
df_joined = df_joined.drop('Postal Code', axis=1)

In [169]:
df_joined.head()

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [170]:
df.shape

(103, 3)

## Map to see different postal codes

In [171]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [177]:
import folium
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_joined['Latitude'], df_joined['Longitude'], df_joined['Borough'], df_joined['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Foursquare Data

In [ ]:
CLIENT_ID = 'id' # your Foursquare ID
CLIENT_SECRET = 'secret' # your Foursquare Secret
VERSION = '20200401' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

In [181]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Create new df with torontos venues

In [182]:
toronto_venues = getNearbyVenues(names=df_joined['Neighborhood'],
                                   latitudes=df_joined['Latitude'],
                                   longitudes=df_joined['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview
The Danforth West, Ri

In [183]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                        Venue  Venue Latitude  Venue Longitude  \
0             Brookbanks Park       43.751976       -79.332140   
1               Variety Store       43.751974       -79.333114   
2  TTC stop - 44 Valley Woods       43.755402       -79.333741   
3      Victoria Village Arena       43.723481       -79.315635   
4                 Tim Hortons       43.725517       -79.313103   

      Venue Category  
0               Park  
1  Food & Drink Shop  
2           Bus Stop  
3       Hockey Arena  
4        Coffee Shop

In [184]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  9   
Bathurst Manor, Wilson Heights, Downsview North                        20   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            56   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing CentrE                                  15   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     77   
Christie                                                               18   
Church and Wellesley                                                   81   
Clarks Corners, Tam O'Shanter, Sullivan                                15   
Cliffside, Cliffcrest, Scarborough Village West                         3   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills                                                              25   
Dorset Park, Wexford Heights, Scarborough Town ...                      7   
Downsview                                                              13   
Dufferin, Dovercourt Village                                           17   
East Toronto                                                            4   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           66   
First Canadian Place, Underground city                                100   
...                                                                   ...   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                3   
Rouge Hill, Port Union, Highland Creek                                  1   
Runnymede, Swansea                                                     39   
Runnymede, The Junction North                                           4   
Scarborough Village                                                     3   
South Steeles, Silverstone, Humbergate, Jamesto...                      9   
St. James Town                                                        100   
St. James Town, Cabbagetown                                            44   
Steeles West, L'Amoreaux West                                          12   
Stn A PO Boxes                                                         96   
Studio District                                                        42   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
The Kingsway, Montgomery Road , Old Mill North                          2   
Thorncliffe Park                                

In [223]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 267 uniques categories.


## One Hot for categories

In [260]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant      ...        Trail  Train Station  \
0                 0                    0      ...            0              0   
1                 0                    0      ...            0              0   
2                 0                    0      ...            0              0   
3                 0                    0      ...            0              0   
4                 0                    0      ...            0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  
0                0         0          0            0              0  
1                0         0          0            0              0  
2                0         0          0            0              0  
3                0         0          0            0              0  
4                0         0          0            0              0  

[5 rows x 267 columns]

In [261]:
toronto_onehot.shape

(2206, 267)

### Get frequency by neighborhood

In [262]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                              Alderwood, Long Branch     0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                   Bedford Park, Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                         Birch Cliff, Cliffside West     0.000000   
7        Brockton, Parkdale Village, Exhibition Place     0.000000   
8               Business reply mail Processing CentrE     0.066667   
9   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
10                                Caledonia-Fairbanks     0.000000   
11              Canada Post Gateway Processing Centre     0.000000   
12                                          Cedarbrae     0.000000   
13                                 Central Bay Street     0.012987   
14                                           Christie     0.000000   
15                               Church and Wellesley     0.024691   
16            Clarks Corners, Tam O'Shanter, Sullivan     0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West     0.000000   
18                     Commerce Court, Victoria Hotel     0.000000   
19                                         Davisville     0.000000   
20                                   Davisville North     0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
22                                          Don Mills     0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...     0.000000   
24                                          Downsview     0.000000   
25                       Dufferin, Dovercourt Village     0.000000   
26                                       East Toronto     0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
28                       Fairview, Henry Farm, Oriole     0.000000   
29             First Canadian Place, Underground city     0.000000   
..                                                ...          ...   
63                          Regent Park, Harbourfront     0.022222   
64                           Richmond, Adelaide, King     0.000000   
65                                           Rosedale     0.000000   
66                                           Roselawn     0.000000   
67             Rouge Hill, Port Union, Highland Creek     0.000000   
68                                 Runnymede, Swansea     0.025641   
69                      Runnymede, The Junction North     0.000000   
70                                Scarborough Village     0.000000   
71  South Steeles, Silverstone, Humbergate, Jamest...     0.000000   
72                                     St. James Town     0.000000   
73                        St. James Town, Cabbagetown     0.000000   
74                      Steeles West, L'Amoreaux West     0.000000   
75                                     Stn A PO Boxes     0.000000   
76                                    Studio District     0.023810   
77  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
78                The Annex, North Midtown, Yorkville     0.000000   
79                                        The Beaches     0.000000   
80                       The Danforth West, Riverdale     0.023810   
81     The Kingsway, Montgomery Road , Old Mill North     0.000000   
82                                   Thorncliffe Park     0.058824   
83           Toronto Dominion Centre, Design Exchange     0.000000   
84                     University of Toronto, Harbord     0.028571   
85                                   Victoria Village     0.000000   
86                                          Westmount     0.000000   
87                                             Weston     0.000000   
88                                  Wexford, Maryva

In [263]:
toronto_grouped.shape

(93, 267)

### Put most common venues per group in DF

In [264]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [265]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot        Clothing Store                Lounge   
1           Coffee Shop              Pharmacy        Sandwich Place   
2           Coffee Shop     Convenience Store        Ice Cream Shop   
3                  Bank   Japanese Restaurant    Chinese Restaurant   
4           Coffee Shop    Italian Restaurant            Restaurant   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant     Electronics Store   Dumpling Restaurant   
1                 Skating Rink    Athletics & Sports                   Pub   
2                  Supermarket         Deli / Bodega      Sushi Restaurant   
3                      Dog Run    Dim Sum Restaurant                 Diner   
4                Grocery Store       Thai Restaurant                   Pub   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0   Empanada Restaurant               Dessert Shop              Drugstore  
1                   Gym                 Comic Shop          Deli / Bodega  
2            Restaurant  Middle Eastern Restaurant                  Diner  
3        Discount Store        Distribution Center       Doner Restaurant  
4                  Café           Sushi Restaurant      Indian Restaurant

## Cluster Neighborhoods

In [266]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 3

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

### Add cluster label to most popular venues DF

In [267]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_joined

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             2.0                  Park   
1  43.725882 -79.315572             0.0          Intersection   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bus Stop     Food & Drink Shop   Distribution Center   
1             Coffee Shop           Pizza Place          Hockey Arena   
2                    Park                   Pub                Bakery   
3  Furniture / Home Store         Women's Store           Coffee Shop   
4             Yoga Studio   Distribution Center    Mexican Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Dessert Shop    Dim Sum Restaurant                 Diner   
1  Portuguese Restaurant         Women's Store          Dessert Shop   
2                Theater            Restaurant        Breakfast Spot   
3     Miscellaneous Shop              Boutique   Sporting Goods Shop   
4                   Bank                   Bar                 Diner   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0        Discount Store                Dog Run   Ethiopian Restaurant  
1    Dim Sum Restaurant                  Diner         Discount Store  
2                  Café     Mexican Restaurant           Dessert Shop  
3           Event Space  Vietnamese Restaurant              Gift Shop  
4        Discount Store               Beer Bar             Hobby Shop

In [268]:
print(set(toronto_merged['Cluster Labels']))
toronto_merged[['Cluster Labels']] = toronto_merged[['Cluster Labels']].fillna(value=3)
toronto_merged.astype({'Cluster Labels': 'int32'})

{0.0, nan, 2.0, 1.0, nan, nan, nan, nan}


Postal code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                            Regent Park, Harbourfront  43.654260 -79.360636   
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
5                                     Islington Avenue  43.667856 -79.532242   
6                                       Malvern, Rouge  43.806686 -79.194353   
7                                            Don Mills  43.745906 -79.352188   
8                      Parkview Hill, Woodbine Gardens  43.706397 -79.309937   
9                             Garden District, Ryerson  43.657162 -79.378937   
10                                           Glencairn  43.709577 -79.445073   
11   West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724   
12              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
13                                           Don Mills  43.725900 -79.340923   
14                                    Woodbine Heights  43.695344 -79.318389   
15                                      St. James Town  43.651494 -79.375418   
16                                  Humewood-Cedarvale  43.693781 -79.428191   
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201   
18                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
19                                         The Beaches  43.676357 -79.293031   
20                           

In [269]:
toronto_merged = toronto_merged.astype({'Cluster Labels': 'int32'})
print(set(toronto_merged['Cluster Labels']))

{0, 1, 2, 3}


### Visualize Clusters

In [270]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters+1)
ys = [i + x + (i*x)**2 for i in range(kclusters+1)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster],
        fill=True,
        fill_color=rainbow[cluster],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Conclusion

We can observe that the 3 clusters are not very close geographically. This means that independantly from the location, different neighborhoods have similar venues around. This shows that you can find a neighborhood similar to yours in any part of Toronto.